File: autoviz/examples/python/16_tutorial_parameters.py
#!/usr/bin/env python3
"""Host an Autolink ParameterServer for Autoviz Parameters panel.

  Node: /autoviz/params
  Params: max_speed (double), enable_estop (bool), waypoint_count (int),
          robot_name (string), goal_tol (double)

Parameters panel: select Node=/autoviz/params → view / edit values.

  /usr/bin/python3 examples/python/16_tutorial_parameters.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink

NODE = '/autoviz/params'

SEED = (
    autolink.Parameter('max_speed', 1.5),
    autolink.Parameter('goal_tol', 0.05),
    autolink.Parameter('enable_estop', False),
    autolink.Parameter('waypoint_count', 8),
    autolink.Parameter('robot_name', 'demo_bot'),
)


def main():
    autolink.init('parameters')
    n = autolink.Node(NODE)
    srv = autolink.ParameterServer(n)
    time.sleep(0.3)
    for p in SEED:
        srv.set_parameter(p)

    listed = srv.list_parameters()
    print(f'ParameterServer {NODE} ({len(listed)} params):')
    for p in listed:
        print(f'  {p.name()}: {p.debug_string()}')
    print('Open Autoviz Parameters → select this node → edit Value')

    try:
        while not autolink.is_shutdown():
            time.sleep(0.2)
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()

File: autoviz/examples/cpp/16_tutorial_parameters.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <chrono>
#include <iostream>
#include <memory>
#include <thread>
#include <vector>

#include "autolink/parameter/parameter.hpp"
#include "autolink/parameter/parameter_server.hpp"

#include "common/tutorial_utils.hpp"

int main(int argc, char** argv) {
  (void)argc;
  if (!autolink::Init(argv[0])) return 1;

  auto node = autolink::CreateNode("/autoviz/params");
  auto srv = std::make_shared<autolink::ParameterServer>(node);
  std::this_thread::sleep_for(std::chrono::milliseconds(300));

  srv->SetParameter(autolink::Parameter("max_speed", 1.5));
  srv->SetParameter(autolink::Parameter("goal_tol", 0.05));
  srv->SetParameter(autolink::Parameter("enable_estop", false));
  srv->SetParameter(autolink::Parameter("waypoint_count", 8));
  srv->SetParameter(autolink::Parameter("robot_name", std::string("demo_bot")));

  std::vector<autolink::Parameter> listed;
  srv->ListParameters(&listed);
  std::cout << "ParameterServer /autoviz/params (" << listed.size()
            << " params):\n";
  for (const auto& p : listed) {
    std::cout << "  " << p.Name() << ": " << p.DebugString() << "\n";
  }
  std::cout << "Open Autoviz Parameters → select this node → edit Value\n";

  while (autolink::OK()) {
    std::this_thread::sleep_for(std::chrono::milliseconds(200));
  }
  return 0;
}
